Mapping System for AKD Framework

muthukumaranR opened the pull request:
#  Mapping System for AKD Framework

## Description

### Summary

This PR introduces a **type-safe mapping system** that enables data transformation between heterogeneous agent schemas in multi-agent workflows. The mapper serves as the critical "glue" that allows agents with incompatible input/output schemas to work together in LangGraph-orchestrated workflows, solving the problem of agent interoperability in the AKD framework.

### Details

1. **Implemented WaterfallMapper with 4-stage progressive strategy**:
   - Direct field matching for exact name correspondence
   - Semantic field matching using fuzzy logic and domain knowledge
   - LLM-assisted parsing for complex transformations
   - String fallback for graceful degradation

2. **Created type-safe mapping infrastructure**:
   - `MappingInput`/`MappingOutput` schemas for full type safety
   - Pydantic model instance handling with schema introspection
   - Integration with AKDSerializer for proper model conversions

3. **Built comprehensive strategy implementations**:
   - `DirectFieldMapper` for fast exact field matching
   - `SemanticFieldMapper` with configurable similarity thresholds
   - `LLMFallbackMapper` with OpenAI integration and graceful fallback

4. **Added performance and reliability features**:
   - Caching system for repeated mappings
   - Circuit breaker pattern to disable failing strategies
   - Confidence scoring for mapping quality assessment
   - Parallel processing for semantic matching

5. **Implemented comprehensive testing**:
   - Unit tests for each mapping strategy
   - Integration tests with real AKD agent schemas
   - Performance and error handling test scenarios
   - Complete agent pipeline workflow tests

6. **Created extensive documentation**:
   - Comprehensive README with architecture diagrams
   - Usage examples and integration patterns
   - LangGraph workflow integration examples
   - Performance considerations and troubleshooting guide


### Files Added

- `akd/mapping/mappers.py` - Core mapping implementation
- `akd/mapping/README.md` - Comprehensive documentation and examples
- `tests/mapping/test_mappers.py` - Full test suite

### Usage Example

```python
from akd.mapping.mappers import WaterfallMapper, MappingInput
from akd.agents.query import QueryAgentOutputSchema
from akd.agents.litsearch import LitAgentInputSchema

# Transform between incompatible agent schemas
mapper = WaterfallMapper()
result = await mapper.arun(MappingInput(
    source_model=query_output,
    target_schema=LitAgentInputSchema,
    mapping_hints={"queries": "query"}  # Optional hints
))

# Access results with confidence scoring
print(f"Confidence: {result.mapping_confidence}")
print(f"Strategy: {result.used_strategy}")
mapped_agent_input = result.mapped_model
```


---

**For comprehensive documentation, architecture diagrams, and advanced examples, see [`akd/mapping/README.md`](./akd/mapping/README.md)**

muthukumaranR committed:
Add AKD Mapping Module

- Introduced a new module for data transformation and mapping within the AKD framework.
- Implemented a waterfall mapping approach with various mapping strategies including Direct, Semantic, and LLM-based mapping.
- Defined input and output schemas for mapping operations.

muthukumaranR committed:
Add BaseMappingStrategy

- Introduced a new mapping module in the AKD framework, implementing a waterfall approach for data transformation.
- Defined MappingConfig for configuration settings, MappingInput for type-safe input specifications, and MappingOutput for transformation results.
- Established an abstract base class, BaseMappingStrategy, to standardize mapping strategies with error handling and validation.
- Enhanced the system with detailed documentation and type annotations for improved clarity.

muthukumaranR committed:
Add DirectFieldMapper, SemanticFieldMapper, and LLMFallbackMapper

- Implemented DirectFieldMapper for exact field name matching, enhancing performance and reliability in data mapping.
- Introduced SemanticFieldMapper to utilize fuzzy semantic matching, accommodating variations in naming conventions.
- Added LLMFallbackMapper to leverage language models for complex data transformations when other strategies fail.
- Enhanced WaterfallMapper to orchestrate these strategies, ensuring robust data compatibility and progressive mapping attempts.
- Included comprehensive documentation and type annotations for clarity and maintainability.

muthukumaranR committed:
Add README for AKD Mapping System

- Introduced comprehensive documentation for the AKD Mapping System, detailing its purpose, architecture, and usage.
- Explained the problem of heterogeneous multi-agent communication and the solution provided by the mapping system.
- Described the waterfall mapping strategy, including stages of direct, semantic, structure-based matching, and LLM fallback.
- Included usage examples, configuration options, error handling, and advanced features to guide users in implementing the mapping system effectively.
- Enhanced overall clarity and maintainability of the documentation.

muthukumaranR committed:
Add comprehensive tests pytest() for AKD mapping system

- Introduced a new test module for the AKD mapping system, validating the functionality of various mappers including DirectFieldMapper, SemanticFieldMapper, and LLMFallbackMapper.
- Implemented tests for akd agent schemas, ensuring proper data transformation between different agent types.
- Included tests for direct field matching, semantic similarity, and LLM fallback behavior, enhancing the reliability of the mapping strategies.
- Added error handling and edge case tests to ensure robustness in schema compatibility and mapping confidence.

NISH1001 commented on akd/mapping/mappers.py:
```diff
@@ -0,0 +1,807 @@
+    target_schema: Type[BaseModel] = Field(
+        description="Target schema class for type-aware mapping"
+    )
+    mapping_hints: Optional[Dict[str, str]] = Field(
```
In python 3.12 + we can directly use `dict[str, str] | None` (just a FYI), not necessary for the review. haha

NISH1001 commented on akd/mapping/mappers.py:
```diff
@@ -0,0 +1,807 @@
+from akd.serializers import AKDSerializer
+
+
+class MappingConfig(BaseConfig):
```
Could this be `MapperInput` ?

NISH1001 commented on akd/mapping/mappers.py:
```diff
@@ -0,0 +1,807 @@
+    )
+
+
+class MappingOutput(OutputSchema):
```
Could this be `MapperOutput`?

NISH1001 commented on akd/mapping/mappers.py:
```diff
@@ -0,0 +1,807 @@
+    def __init__(self, config: Optional[MappingConfig] = None):
+        super().__init__(config=config or MappingConfig())
+        self.failure_count = 0
+        self.is_disabled = False
```
Nitpick: Could we add a one-line comment for what this `is_disabeld` does?

NISH1001 commented on akd/mapping/mappers.py:
```diff
@@ -0,0 +1,807 @@
+            "counts": ["count", "total", "length", "size", "num_results"],
+            "urls": ["url", "link", "href", "uri", "source"],
+        }
+
```
Can we have this dictionary mapping in the config for semantic field mapper? And default to this?

Or just `SemanticFieldMapper._SEMANTIC_GROUPS`

That will provide some control/dynamism to change it externally.

NISH1001 commented on akd/mapping/mappers.py:
```diff
@@ -0,0 +1,807 @@
+6. Be conservative but creative in your mappings
+
+Return only the JSON object, no additional text:
+"""
```
have this as config part as well? Or just `LLMFallbackMapper._PROMPT`? And then we can use `str.format(...)` method to fill in the keys with values.

NISH1001 commented on akd/mapping/mappers.py:
```diff
@@ -0,0 +1,807 @@
+
+        # Cache for performance
+        self._cache: Dict[str, MappingOutput] = {}
+        self.serializer = AKDSerializer()
```
This `serializer` line could be ommitted as the base class has it

NISH1001 commented on akd/mapping/mappers.py:
```diff
@@ -0,0 +1,807 @@
+        # Initialize mapping strategies
+        self.direct_mapper = DirectFieldMapper(self.config)
+        self.semantic_mapper = SemanticFieldMapper(self.config)
+        self.llm_mapper = LLMFallbackMapper(self.config)
```
Can we provide this as dependency injected?
```python
...
def __init__(self, config, *mappers) -> None:
  ...
  self.mappers = mappers
```

This way in future if we have other implementations, we can just pass them as well

NISH1001 commented on akd/mapping/mappers.py:
```diff
@@ -0,0 +1,807 @@
+            ("direct", self.direct_mapper, self.enable_direct_matching),
+            ("semantic", self.semantic_mapper, self.enable_semantic_matching),
+            ("llm_fallback", self.llm_mapper, self.enable_llm_fallback),
+        ]
```
If we have dependency injection, this list could be dynamically created as well.

muthukumaranR committed:
Move LLM transformation prompt to prompts.py

- Introduced a new LLM transformation prompt in `prompts.py` to standardize the data transformation process into target schema formats.

muthukumaranR committed:
Refactor mapping classes
- Renamed MappingInput, MappingOutput, and MappingConfig to MapperInput, MapperOutput, and MapperConfig for consistency across the mapping module.
- Updated tests to reflect the new naming conventions and ensure compatibility with the refactored classes.

muthukumaranR committed:
Add semantic field mapping groups in MapperConfig

muthukumaranR committed:
Refactor WaterfallMapper to support dynamic mapper creation

- Updated WaterfallMapper to initialize mappers based on configuration flags, enhancing flexibility in mapping strategy selection.
- Removed hardcoded strategy definitions in favor of a dynamic approach that instantiates only enabled mappers.
- Added error handling for disabled mappers and improved logging for failed mapping attempts.
- Introduced a new private method `_create_default_mappers` to encapsulate the logic for creating default mappers based on configuration.

muthukumaranR committed:
Enhance MapperInput type definition and add Circuit Breaker flag

- Updated the type definition of `mapping_hints` in `MapperInput` to use `dict[str, str] | None` for improved type safety.
- Introduced a Circuit Breaker flag in `BaseMappingStrategy` to disable the mapping strategy after repeated failures, enhancing error handling.
- Removed redundant confidence check in `LLMFallbackMapper` to streamline the estimation process.

muthukumaranR committed:
Update README to reflect renaming of MappingInput to MapperInput

- Changed all instances of `MappingInput` to `MapperInput` in the README documentation to maintain consistency with recent refactoring.
- Ensured that examples and code snippets are updated accordingly for clarity and accuracy.

NISH1001 commented on akd/mapping/mappers.py:
```diff
@@ -7,18 +7,25 @@
 
+from langchain_openai import (
+    ChatOpenAI,
+)  # TODO: implement liteLLM interface and replace this
```
Ya. TODO for whole akd as well.

NISH1001 commented on akd/mapping/mappers.py:
```diff
@@ -304,20 +327,11 @@ class SemanticFieldMapper(BaseMappingStrategy):
-            "urls": ["url", "link", "href", "uri", "source"],
-        }
+        # Use semantic groups from config
+        self.semantic_groups = self.config.semantic_groups
```
BTW, the base class automaitcally sets the attributes dynamically (at runtime) from config. So, we can avoid doing this as well. And it avoids rewriting the same constructor pattern like the base class has at `AbstractBase`. Just small FYI. :)

See [here](https://github.com/NASA-IMPACT/accelerated-discovery/blob/5e502487eb7d8684a687671df9622fac24b65a9d/akd/_base.py#L149) `__set_attrs_from_config` and this is within `_post_init` which is called within the constructor.

muthukumaranR replied:
Ah okay, didn't see the `__set_attrs_from_config`. Thanks for the catch

NISH1001 commented on akd/mapping/mappers.py:
```diff
@@ -714,19 +727,15 @@ async def _arun(self, params: MappingInput, **kwargs) -> MappingOutput:
 
-        for strategy_name, mapper, enabled in strategies:
-            if not enabled or mapper.is_disabled:
+        for mapper in self.mappers:
+            # Skip disabled mappers (circuit breaker)
+            if mapper.is_disabled:
```
Nice. Thanks for the changes. I like this better.

muthukumaranR committed:
- Removed redundant assignment of semantic_groups in `SemanticFieldMapper`

NISH1001 reviewed (approved):
This looks cleaner now. Thanks for addressing the comments.
